HWKAPM-567 - Removed waits for a fixed amount of time by waits on conditions.

jpkrohling opened the pull request.

jpkrohling commented:
@objectiser , @pavolloffay , would you please review? It would also be good to run the tests locally: `mvn clean install -Ptests -Dmaven.test.failure.ignore=true` 

On master, I had to try a few times to get a build to finish, as it would block "somewhere". With this change, I get test failures, but the build finishes.

A good side effect is that the build is now faster. 

Before:
```
[INFO] Hawkular APM::Tests::Instrumentation ............... SUCCESS [04:02 min]
[INFO] Hawkular APM::Tests::Containers::Standalone ........ SUCCESS [  9.102 s]
[INFO] Hawkular APM::Tests::Containers::Jetty ............. SUCCESS [02:18 min]
[INFO] Hawkular APM::Tests::Containers::Vertx ............. SUCCESS [ 38.723 s]
[INFO] Hawkular APM::Tests::Containers::Wildfly ........... SUCCESS [ 56.693 s]
[INFO] Hawkular APM::Tests::Server::Wildfly ............... SUCCESS [02:22 min]
[INFO] Hawkular APM::Tests::Tools::Instrumenter ........... SUCCESS [  6.848 s]
[INFO] Hawkular APM::Tests::Instrumentation Test Framework  SUCCESS [  3.693 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 10:39 min
[INFO] Finished at: 2016-08-11T18:43:40+02:00
[INFO] Final Memory: 104M/726M
[INFO] ------------------------------------------------------------------------
```

After:
```
[INFO] Hawkular APM::Tests::Instrumentation ............... SUCCESS [ 55.268 s]
[INFO] Hawkular APM::Tests::Containers::Standalone ........ SUCCESS [ 14.021 s]
[INFO] Hawkular APM::Tests::Containers::Jetty ............. SUCCESS [ 21.186 s]
[INFO] Hawkular APM::Tests::Containers::Vertx ............. SUCCESS [ 37.877 s]
[INFO] Hawkular APM::Tests::Containers::Wildfly ........... SUCCESS [ 51.838 s]
[INFO] Hawkular APM::Tests::Server::Wildfly ............... SUCCESS [01:27 min]
[INFO] Hawkular APM::Tests::Tools::Instrumenter ........... SUCCESS [  3.847 s]
[INFO] Hawkular APM::Tests::Instrumentation Test Framework  SUCCESS [  3.842 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 04:44 min
[INFO] Finished at: 2016-08-12T14:39:36+02:00
[INFO] Final Memory: 123M/793M
[INFO] ------------------------------------------------------------------------
```

These times are taken from a "-rf" build, meaning, it starts at the `tests/commons` module.

objectiser commented:
I'll take a look shortly.....

objectiser commented on tests/common/src/main/java/org/hawkular/apm/tests/common/Wait.java:
```diff
@@ -0,0 +1,123 @@
+     * @param timeUnit the unit for the timeout
+     */
+    public static void until(Callable<Boolean> condition, long timeout, TimeUnit timeUnit) {
+        until(condition, 10, TimeUnit.SECONDS, 50);
```
Shouldn't this be passing through the parameters 'timeout' and 'timeUnit'?

jpkrohling replied:
+1

pavolloffay commented:
`hawkular-apm <PR517>$ mvn clean install -Ptests -Dmaven.test.failure.ignore=true >> ~/out.txt 2>&1`

Here is full log:
[out.txt](https://github.com/hawkular/hawkular-apm/files/415576/out.txt)

I get lot of exceptions `java.lang.AssertionError: java.net.NoRouteToHostException: No route to host` and test failures.

```
.
.
.
[INFO] Hawkular APM::Server::WAR .......................... SUCCESS [  1.462 s]
[INFO] Hawkular APM::Tools::Instrumenter .................. SUCCESS [  1.865 s]
[INFO] Hawkular APM::UI::Console .......................... SUCCESS [02:20 min]
[INFO] Hawkular APM::Distribution ......................... SUCCESS [ 29.060 s]
[INFO] Hawkular APM::Tests::Server::Mock .................. SUCCESS [  0.523 s]
[INFO] Hawkular APM::Tests::Common ........................ SUCCESS [  0.292 s]
[INFO] Hawkular APM::Tests::Instrumentation ............... SUCCESS [ 53.584 s]
[INFO] Hawkular APM::Tests::Containers::Standalone ........ SUCCESS [  3.075 s]
[INFO] Hawkular APM::Tests::Containers::Jetty ............. SUCCESS [ 19.446 s]
[INFO] Hawkular APM::Tests::Containers::Vertx ............. SUCCESS [01:15 min]
[INFO] Hawkular APM::Tests::Containers::Wildfly ........... SUCCESS [ 37.446 s]
[INFO] Hawkular APM::Tests::Server::Wildfly ............... SUCCESS [ 52.110 s]
[INFO] Hawkular APM::Tests::Tools::Instrumenter ........... SUCCESS [  2.645 s]
[INFO] Hawkular APM::Tests::Instrumentation Test Framework  SUCCESS [  2.295 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 09:40 min
[INFO] Finished at: 2016-08-12T16:55:54+02:00
[INFO] Final Memory: 171M/1488M
[INFO] ------------------------------------------------------------------------
```

objectiser commented on tests/container/vertx/src/test/java/org/hawkular/apm/tests/client/vertx/ClientVertxEventBusTest.java:
```diff
@@ -72,14 +73,6 @@ public void handle(AsyncResult<Message<String>> message) {
-            e.printStackTrace();
-        }
-
         mc.unregister();
```
Haven't run the build yet, so may work fine - but just wondering whether if the consumer is unregistered before the completion handler fires (async), then no message would be sent on the bus. I think that is why I had to introduce the delays.

Maybe some synchronization could be performed, to wait until the message has been sent and reply received?

jpkrohling replied:
Forgot to reply to this one (valid also for the next): one solution could be to introduce a CountDownLatch in a second consumer. This second consumer does the countDown and releases the latch. The latch would wait before the "unregister" call. 

Assuming we can register a consumer on the test itself.

objectiser commented on tests/container/vertx/src/test/java/org/hawkular/apm/tests/client/vertx/ClientVertxHTTPTest.java:
```diff
@@ -124,13 +125,7 @@ public void testHTTPPOST() {
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
+        Wait.until(() -> getApmMockServer().getTraces().size() == 2);
```
Same as above.

objectiser commented:
Finished review of code - only the three comments. Haven't tried running it yet, but looked at Pavol's log - I think we need to make sure it runs more reliably before merging - hopefully just a case of identifying the areas where it tends to fail and adjusting the wait conditions.

jpkrohling commented:
I'm working on improving the tests, but wanted to make sure: the integration tests run reliably for both of you, right? For me, it sometimes freezes as if it were waiting for something, and "never" unblocks.

objectiser commented:
@jpkrohling will try now and let you know

objectiser commented:
@jpkrohling Build ran to completion for me, with the maven.test.failure.ignore set to true.

jpkrohling commented:
This PR has been updated to fix a few tests. The only one that I still have problems with is the `NettyHttpTest`  in `tests/instrumentation`.

jpkrohling commented:
I updated the PR to remove more waits. The overall stability is good, and the only problem I have is with the Netty tests, which I was having already. As mentioned on IRC, I switched from blocking indefinitely to blocking for 5 seconds at most on those tests.

@objectiser , @pavolloffay , would you please re-run the test suite and let me know the results? 

`mvn clean install -Ptests -Dmaven.test.failure.ignore=true > /tmp/maven.log`

jpkrohling commented:
On a full local run, I did see some unexpected failures.

objectiser commented:
@jpkrohling Could you rebase and then I'll try again.

objectiser commented:
@jpkrohling I had some exceptions in the TraceServiceRESTTest

SEVERE: [TraceServiceRESTClient] [Thread[pool-16-thread-1,5,main]] Failed to send 'get' trace request
java.io.IOException: Server returned HTTP response code: 500 for URL: http://127.0.0.1:8080/hawkular/apm/traces/complete/1
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1840)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1441)
	at org.hawkular.apm.trace.service.rest.client.TraceServiceRESTClient.getTrace(TraceServiceRESTClient.java:144)
	at org.hawkular.apm.tests.wildfly.TraceServiceRESTTest.lambda$testStoreAndRetrieveComplexTraceById$2(TraceServiceRESTTest.java:178)
	at org.hawkular.apm.tests.common.Wait.lambda$until$0(Wait.java:75)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

pavolloffay commented:
Here is my log:

[maven.txt](https://github.com/hawkular/hawkular-apm/files/424635/maven.txt)

jpkrohling commented:
@pavolloffay , your fails only on the Netty client, which is "acceptable". I'll create a JIRA to investigate better why this fails with the Wait and not with the old wait. Perhaps we are facing a deadlock somewhere that might uncover a bug.

About the failure from @objectiser , I'll change this specific condition checker to accept exceptions and keep trying, but ideally, it wouldn't throw a 500 for an unknown entity. The same test passes for @pavolloffay , so, I think the 500 happens because the record doesn't exist at the time of the first request.

jpkrohling commented:
https://issues.jboss.org/browse/HWKAPM-584 - for the 500 that @objectiser reported.

jpkrohling commented:
I believe this is now ready for another round of reviews and tests. If everyone is satisfied, it can be merged as well.

objectiser commented:
Tests ran to completion - although the same TraceServiceRESTTest exceptions were seen in the console, there was no evidence of anything in the server.log. I've added both information to the jira - and it didn't cause any test failures.

jpkrohling commented:
retest this please

jpkrohling commented:
Is there anything missing on this PR, or could this be merged?

objectiser commented:
@jpkrohling Just rerunning travis to see if can get green build - will then merge.

jpkrohling commented:
retest this please

jpkrohling commented:
(the above is a comment to Jenkins, to try this PR again as well)

objectiser commented:
@jpkrohling Thanks Juca!
